[ADD] estate:  create the Estate Module

shsak-odoo opened the pull request:
1. estate Module:
- Creates various models to manage Estate Property, Tags, Types, Offers
- Creates menus and actions for better UI
- Sets onchange and action listeners to change the values based on events.
- Adds validation to various fields to avoid unwanted results, like a strictly positive value for the expected and selling price of the property.
- Handles UserError and ValidationError
- add invoicing using the new `estate_account` link module
2. awesome_owl Module:
- Display counter and manage its state using the useState hook
- Create a sub-component for the counter
- props validation
- add TODO list
- auto focus on input
- deleting TODO
- generic card using a slot to view content
3. awesome_dashboard Module:
- add button to navigate to CRM Leads and Customers model
- make the network call to fetch the statistics
- cache network  call using memoize service
- display a pie chart
- real-life network call-in every 10 minutes
- lazy loading of the dashboard so that it only loads when needed
- make the dashboard generic
- make the dashboard extensible
- can add or remove the dashboard items, preferences are stored in the local storage

shsak-odoo committed:
[ADD] estate:  added init file and manifest file to make it a module

Created basic module structure

Chapter 2

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/tutorials/pull/1047.png)](https://mergebot.odoo.com/odoo/tutorials/pull/1047)

shsak-odoo committed:
[IMP] estate:  Create estate property model with more field set

Adds the main estate property model and includes a broad set of fields
to better describe each property.

Chapter 3

shsak-odoo committed:
[IMP] estate:  Grant base users full access to estate property by security rules

Provided the base user group with all the permission for the Estate.property model.

Chapter 4

shsak-odoo committed:
[IMP] estate:  Added menus,  actions, and views for the UI

Added menus to show the application on the odoo root menu
Action to open the view
List and form view to Create, Edit, and Show Estate Property

Chapter 5

shsak-odoo committed:
[IMP] estate:  Added form, search, filters, groups views

Allows users to search across multiple fields
Enables filtered results and grouping by various parameters

Chapter 6

shsak-odoo committed:
[IMP] estate:  Implement relations between modules

Added relational fields

Chapter 7

shsak-odoo committed:
[IMP] estate:  Add compute fields and onchanges for offer and garden

Added compute logic for offer validity and deadline_date
Implemented onchange methods to set default garden area values

Chapter 8

bit-odoo commented on estate/__manifest__.py:
```diff
@@ -0,0 +1,20 @@
+{
+    'name': "Real Estate",
+    'version': '1.0',
```
Can you please follow the same quote in the same file?

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,60 @@
+    garden_area = fields.Integer()
+    garden_orientation = fields.Selection(
+        string='Garden Orientation',
+        selection=[('north', 'North'), ('south', 'South'), ('east', 'East'), ('west', 'West')]
```
We have to follow the same quote in the same file.

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,60 @@
+            self.garden_orientation = 'north'
+            self.garden_area = 10
+        else:
+            self.garden_orientation = ''
```
```suggestion
            self.garden_orientation = None
```

bit-odoo commented on estate/security/ir.model.access.csv:
```diff
@@ -0,0 +1,5 @@
+estate.access_estate_property,access_estate_property,estate.model_estate_property,base.group_user,1,1,1,1
+estate.access_estate_property_type,access_estate_property_type,estate.model_estate_property_type,base.group_user,1,1,1,1
+estate.access_estate_property_tag,access_estate_property_tag,estate.model_estate_property_tag,base.group_user,1,1,1,1
+estate.access_estate_property_offer,access_estate_property_offer,estate.model_estate_property_offer,base.group_user,1,1,1,1
```
There should be one empty line at EOF.

bit-odoo commented on estate/views/estate_menus.xml:
```diff
@@ -0,0 +1,13 @@
+                action="estate_property_tag_action" />
+        </menuitem>
+    </menuitem>
+</odoo>
```
There should be one empty line at EOF.

bit-odoo commented on estate/views/estate_menus.xml:
```diff
@@ -0,0 +1,13 @@
+        <menuitem id="estate_first_level_menu" name="Advertisement">
+            <menuitem id="estate_property_model_menu_action" action="estate_property_action" />
+        </menuitem>
+        <menuitem id="estate_first_level_setting_menu" name="Settings">
```
```suggestion
        
        <menuitem id="estate_first_level_setting_menu" name="Settings">
```
For better visibility.

bit-odoo commented on estate/views/estate_property_offer_views.xml:
```diff
@@ -0,0 +1,33 @@
+            </form>
+        </field>
+    </record>
+</odoo>
```
There should be one empty line at EOF.

bit-odoo commented on estate/views/estate_property_tag_views.xml:
```diff
@@ -0,0 +1,7 @@
+        <field name="res_model">estate.property.tag</field>
+        <field name="view_mode">list,form</field>
+    </record>
+</odoo>
```
There should be one empty line at EOF.

bit-odoo commented on estate/views/estate_property_type_views.xml:
```diff
@@ -0,0 +1,7 @@
+        <field name="res_model">estate.property.type</field>
+        <field name="view_mode">list,form</field>
+    </record>
+</odoo>
```
There should be one empty line at EOF.

bit-odoo commented on estate/views/estate_property_views.xml:
```diff
@@ -0,0 +1,97 @@
+        <field name="res_model">estate_property</field>
+        <field name="view_mode">list,form</field>
+    </record>
+</odoo>
```
There should be one empty line at EOF.

bit-odoo reviewed (commented):
Hello @shsak-odoo 
You did a good job :clap: 

I have added some suggestions. And can you please improve your commit message, and pr message, and title regarding these guidelines - https://www.odoo.com/documentation/19.0/contributing/development/git_guidelines.html

Thanks

shsak-odoo commented:
I have applied the changes based on the review and suggestions.

shsak-odoo committed:
[IMP] estate:  Added Action button in Property and Offer Model

In estate.property model added the action button
1. to cancel the property advertisement
2. to mark the property sold

In estate.offer model added the action button
1. to accept the offer for the particular property
2. to refuse the offer for the particular property
3. set buyer and selling price, when any offer is accepted

Chapter 9

shsak-odoo committed:
[IMP] estate:  Apply constrain to form fields

Added a constraint on the expected price and the selling price
so that they can't be negative

Applied a constraint on the offer's price so that it can't be negative

Added a unique name check for tags and types model

Chapter 10

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,95 @@
+    )
+    state = fields.Selection(
+        string='Status',
+        selection=[('new', 'New'), ('offer_received', 'Offer Received'), ('offer_accepted', 'Offer Accepted'), ('sold', 'Sold'), ('cancelled', 'Cancelled')],
```
```suggestion
        selection=[
            ('new', 'New'),
            ('offer_received', 'Offer Received'),
            ('offer_accepted', 'Offer Accepted'),
            ('sold', 'Sold'),
            ('cancelled', 'Cancelled')
        ],
```

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,95 @@
+                raise exceptions.UserError('A sold property cannot be cancelled')
+            else:
+                record.state = 'cancelled'
+        return True
```
I prefer to start action names with `action_` for easy identification.

It's good to make the error message translatable.
`raise exceptions.UserError(_('Sold property cannot be cancelled'))`

We can use filtered() here.

`if self.filtered(lamda x: x.state=="sold")`

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,95 @@
+                raise exceptions.UserError('A cancelled property cannot be Sold')
+            else:
+                record.state = 'sold'
+        return True
```
Same.

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,95 @@
+    _check_selling_price = models.Constraint(
+        'CHECK(selling_price >= 0)',
+        'The Selling price cannot be less then 0'
+    )
```
It is good if we declare constraints immediately after all field declarations. For better visibility.

bit-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,48 @@
+    def action_offer_accepted(self):
+        for record in self:
+            if record.property_id.buyer:
+                raise exceptions.UserError('An another offer is already accepted')
```
Here also.

bit-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,48 @@
+            record.status = 'refused'
+        return True
+
+    _check_offer_price = models.Constraint(
```
same about declaration :)

bit-odoo commented on estate/views/estate_property_views.xml:
```diff
@@ -0,0 +1,101 @@
+        <field name="arch" type="xml">
+            <form string="Description">
+                <header>
+                    <button type="object" name="cancel_property_sale" string="Cancel" />
```
```suggestion
                    <button type="object" name="action_property_sale" string="Cancel" />
```

bit-odoo commented on estate/views/estate_property_views.xml:
```diff
@@ -0,0 +1,101 @@
+            <form string="Description">
+                <header>
+                    <button type="object" name="cancel_property_sale" string="Cancel" />
+                    <button type="object" name="set_property_sold" string="Sold" />
```
```suggestion
                    <button type="object" name="action_property_sold" string="Sold" />
```

bit-odoo reviewed (commented):
Hello, Good Work!

I have made some suggestions.
Can you please update PR title?

shsak-odoo committed:
[IMP] estate:  Adds inline views, widget, decorations, Stat button

will be able to edit the property tags and offer from inline
without opening the form view

Adds the decoration to properties, offers
Adds a status bar for the property state

Adds a stat button to the property types to view all the
offer related to that property type

shsak-odoo commented:
Hello Sir,
I have applied the changes based on the suggestions.

shsak-odoo committed:
[IMP] estate:  Modify the res.user to add property_ids by inheritance

Overrides ondelete and create method for the models
- on delete for property model
- on create, offer to check if an offer with a higher bid already exists or not

Inherited the res. user to add a new property_id field to it
so that we can list all the properties associated with the User's account.

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,110 @@
+    @api.ondelete(at_uninstall=False)
+    def _unlink_if_new_or_sold(self):
+        if self.filtered(lambda x: x.state not in ('new', 'cancelled')):
+            raise exceptions.UserError(_('Cannot delete property which is not yer sold or cancelled'))
```
TYPO

bit-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,62 @@
+
+    _check_offer_price = models.Constraint(
+        'CHECK(price > 0)',
+        'Offer Price cannot be 0 or below 0'
```
```suggestion
        'The offer price must be greater than 0'
```
I think it is littlebit simpler. What do you think?

shsak-odoo committed:
[IMP] estate:  Create New module estate_account to create Property invoice

Will add a new module that depends on the estate and account module

It overrides the sold action to create an invoice for the property sold

shsak-odoo committed:
[IMP] estate:  add the kanban view for the property list

Properties can now be grouped by type in the Kanban view

Chapter: 14

shsak-odoo committed:
[IMP] awesome_owl:  add counter and card component

- add counter component
- implemented sub-components
- implements callback function
- add markup to card

shsak-odoo committed:
[IMP]  awesome_owl:  add the To-Do components

- can add todo items
- remove the todo items
- mark the to-do items completed
- Any component can be passed as a slot in the card component
- toggles the slot view in the card component

shsak-odoo committed:
[IMP] awesome_dashboard:  add the rpc call to fetch data using services

- add two buttons: Customers and Leads to view them
- use the RPC call to get the statistics data
- create a service to cache the RPC calls data

shsak-odoo committed:
[IMP] awesome_dashboard:  cache the network call to get statistics

- fetch the latest statistics at regular intervals
- lazy loading for the dashboard assets
- displays pie chart

bit-odoo commented on awesome_dashboard/static/src/dashboard/pie_chart/pie_chart.js:
```diff
@@ -0,0 +1,45 @@
+            }
+        }
+    }
+}
```
Should be one empty line at the EOF.

bit-odoo commented on awesome_dashboard/static/src/dashboard/pie_chart/pie_chart.xml:
```diff
@@ -0,0 +1,8 @@
+        <canvas t-ref="canvaRef" />
+    </t>
+
+</templates>
```
Should be one empty line at the EOF.

bit-odoo commented on awesome_dashboard/static/src/dashboard/pie_chart/pie_chart_card.js:
```diff
@@ -0,0 +1,16 @@
+            optional: true
+        },
+    }
+}
```
Should be one empty line at the EOF.

bit-odoo commented on awesome_dashboard/static/src/dashboard/pie_chart/pie_chart_card.xml:
```diff
@@ -0,0 +1,13 @@
+        </DashboardItem>
+    </t>
+
+</templates>
```
Should be one empty line at the EOF.

bit-odoo commented on awesome_dashboard/static/src/dashboard_action.js:
```diff
@@ -0,0 +1,12 @@
+    `;
+}
+
+registry.category("actions").add("awesome_dashboard.dashboard", DashboardComponentLoader);
```
Should be one empty line at the EOF.

bit-odoo commented on awesome_owl/static/src/playground.xml:
```diff
@@ -1,10 +1,32 @@
+            </div>
         </div>
     </t>
-
```
Unneccary diff.

bit-odoo commented on awesome_owl/static/src/utils.js:
```diff
@@ -0,0 +1,13 @@
+    })
+
+    return ref;
+}
```
Should be one empty line at the EOF.

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,118 @@
+
+    name = fields.Char(required=True)
+    description = fields.Text()
+    postcode = fields.Char()
```
Why char ? I think it should be Interger.

bit-odoo commented on estate_account/__manifest__.py:
```diff
@@ -0,0 +1,12 @@
+    A Real Estate Invoicing Module
+    """,
+    "license": "LGPL-3",
+    "data": [],
```
Is it necessary?

bit-odoo reviewed (commented):
Hello,
I have made some suggestions.
And can you please adapt your pr message regarding these guidelines - https://www.odoo.com/documentation/19.0/contributing/development/git_guidelines.html

shsak-odoo committed:
[FIX] awesome_dashboard:  make the dashboard generic and extensible

- dashboard item can have generic props
- dashboard item is loaded through registry
- code spacing and typo fix

shsak-odoo commented:
Hello Sir,
I have made changes based on your suggestions.

shsak-odoo committed:
[IMP] awesome_dashboard:  able to remove and add dashboard items

- Use local storage to store the user preference for the dashboard item to show
-  preference preserved after refresh.
